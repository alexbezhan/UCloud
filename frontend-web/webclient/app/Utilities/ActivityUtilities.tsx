import { buildQueryString } from "./URIUtilities";
import { ScrollRequest } from "Scroll";
import { ActivityFilter } from "Activity";

export const activityQuery = (scroll: ScrollRequest<number>, filter?: ActivityFilter) => {
    let params: any = {};
    if (!!scroll.offset) params.offset = scroll.offset;
    params.scrollSize = scroll.scrollSize;

    if (filter !== undefined) {
        if (filter.type !== undefined) params.type = filter.type;
        if (filter.collapseAt !== undefined) params.collapseAt = filter.collapseAt;
        if (filter.maxTimestamp !== undefined) params.maxTimestamp = filter.maxTimestamp.getTime();
        if (filter.minTimestamp !== undefined) params.minTimestamp = filter.minTimestamp.getTime();
    }

    return buildQueryString("/activity/browse/user", params);
};

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}`;

export const getLatestActivity = () => undefined;