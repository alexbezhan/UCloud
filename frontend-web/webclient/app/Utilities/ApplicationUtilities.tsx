import {
    ApplicationInvocationDescription,
    ApplicationMetadata,
    ApplicationParameter,
    AppState,
    ParameterTypes,
    RunsSortBy
} from "Applications";
import Cloud from "Authentication/lib";
import {SortOrder} from "Files";
import * as React from "react";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {Page} from "Types";
import {addStandardDialog} from "UtilityComponents";
import {errorMessageOrDefault, removeTrailingSlash} from "UtilityFunctions";
import {expandHomeFolder} from "./FileUtilities";

export const hpcJobQueryPost = "/hpc/jobs";

export const hpcJobQuery = (
    id: string,
    stdoutLine: number,
    stderrLine: number,
    stdoutMaxLines: number = 1000,
    stderrMaxLines: number = 1000
) =>
    `/hpc/jobs/follow/${encodeURIComponent(id)}?stdoutLineStart=${stdoutLine}&stdoutMaxLines=${stdoutMaxLines}&stderrLineStart=${stderrLine}&stderrMaxLines=${stderrMaxLines}`;

export function hpcJobsQuery(
    itemsPerPage: number,
    page: number,
    sortOrder?: SortOrder,
    sortBy?: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: AppState
): string {
    let query = `/hpc/jobs/?itemsPerPage=${itemsPerPage}&page=${page}`;
    if (sortOrder) query = query.concat(`&order=${sortOrder}`);
    if (sortBy) query = query.concat(`&sortBy=${sortBy}`);
    if (minTimestamp != null) query = query.concat(`&minTimestamp=${minTimestamp}`);
    if (maxTimestamp != null) query = query.concat(`&maxTimestamp=${maxTimestamp}`);
    if (filter != null) query = query.concat(`&filter=${filter}`);
    return query;
}

export const hpcFavoriteApp = (name: string, version: string) =>
    `/hpc/apps/favorites/${encodeURIComponent(name)}/${encodeURIComponent(version)}`;

export const hpcFavorites = (itemsPerPage: number, pageNumber: number) =>
    `/hpc/apps/favorites?itemsPerPage=${itemsPerPage}&page=${pageNumber}`;

export const hpcApplicationsQuery = (page: number, itemsPerPage: number) =>
    `/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`;

interface HPCApplicationsSearchQuery {
    query: string,
    page: number,
    itemsPerPage: number
}

export const hpcApplicationsSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/search?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const hpcApplicationsTagSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/searchTags?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const cancelJobQuery = `hpc/jobs`;

export const cancelJobDialog = ({jobId, onConfirm, jobCount = 1}: { jobCount?: number, jobId: string, onConfirm: () => void }): void =>
    addStandardDialog({
        title: `Cancel job${jobCount > 1 ? "s" : ""}?`,
        message: jobCount === 1 ? `Cancel job: ${jobId}?` : "Cancel jobs?",
        cancelText: "No",
        confirmText: `Cancel job${jobCount > 1 ? "s" : ""}`,
        onConfirm
    });

export const cancelJob = (cloud: Cloud, jobId: string): Promise<{ request: XMLHttpRequest, response: void }> =>
    cloud.delete(cancelJobQuery, {jobId});

interface FavoriteApplicationFromPage<T> {
    name: string
    version: string
    page: Page<{ metadata: ApplicationMetadata, favorite: boolean } & T>
    cloud: Cloud
}

/**
 * Favorites an application.
 * @param {Application} Application the application to be favorited
 * @param {Cloud} cloud The cloud instance for requests
 */
export async function favoriteApplicationFromPage<T>({name, version, page, cloud}: FavoriteApplicationFromPage<T>): Promise<Page<T>> {
    const a = page.items.find(it => it.metadata.name === name && it.metadata.version === version)!;
    try {
        await cloud.post(hpcFavoriteApp(name, version));
        a.favorite = !a.favorite;
    } catch (e) {
        snackbarStore.addFailure(errorMessageOrDefault(e, `An error ocurred favoriting ${name}`));
    }
    return page;
}


interface StringMap {
    [k: string]: string;
}

interface AllowedParameterKey {
    name: string;
    type: ParameterTypes;
}

interface ExtractParameters {
    nameToValue: StringMap;
    allowedParameterKeys: AllowedParameterKey[];
    siteVersion: number;
}

export const findKnownParameterValues = ({nameToValue, allowedParameterKeys, siteVersion}: ExtractParameters): StringMap => {
    const extractedParameters = {};
    if (siteVersion === 1) {
        allowedParameterKeys.forEach(({name, type}) => {
            if (nameToValue[name] !== undefined) {
                if (typeMatchesValue(type, nameToValue[name])) {
                    extractedParameters[name] = nameToValue[name];
                }
            }
        });
    }
    return extractedParameters;
};

export const isFileOrDirectoryParam = ({type}: { type: string }) => type === "input_file" || type === "input_directory";

const typeMatchesValue = (type: ParameterTypes, parameter: string): boolean => {
    switch (type) {
        case ParameterTypes.Boolean:
            return parameter === "Yes" || parameter === "No" || parameter === "";
        case ParameterTypes.Integer:
            return parseInt(parameter, 10) % 1 === 0;
        case ParameterTypes.FloatingPoint:
            return typeof parseFloat(parameter) === "number";
        case ParameterTypes.Text:
        case ParameterTypes.InputDirectory:
        case ParameterTypes.InputFile:
        case ParameterTypes.SharedFileSystem:
        case ParameterTypes.Peer:
            return typeof parameter === "string";

    }
};

interface ExtractedParameters {
    [key: string]: string | number | boolean |
        { source: string, destination: string; } |
        { fileSystemId: string } |
        { jobId: string }
}

export type ParameterValues = Map<string, React.RefObject<HTMLInputElement | HTMLSelectElement>>;

interface ExtractParametersFromMap {
    map: ParameterValues;
    appParameters: ApplicationParameter[];
    cloud: Cloud;
}

export function extractValuesFromWidgets({map, appParameters, cloud}: ExtractParametersFromMap): ExtractedParameters {
    const extracted: ExtractedParameters = {};
    map.forEach(({current}, key) => {
        const parameter = appParameters.find(it => it.name === key);
        console.log(parameter, current);
        if (!current) return;
        if (!current.value || !current.checkValidity()) return;
        if (!parameter) return;
        switch (parameter.type) {
            case ParameterTypes.InputDirectory:
            case ParameterTypes.InputFile:
                const expandedValue = expandHomeFolder(current.value, cloud.homeFolder);
                extracted[key] = {
                    source: expandedValue,
                    destination: removeTrailingSlash(expandedValue).split("/").pop()!
                };
                return;
            case ParameterTypes.Boolean:
                switch (current.value) {
                    case "Yes":
                        extracted[key] = true;
                        return;
                    case "No":
                        extracted[key] = false;
                        return;
                    default:
                        return;
                }
            case ParameterTypes.Integer:
                extracted[key] = parseInt(current.value, 10);
                return;
            case ParameterTypes.FloatingPoint:
                extracted[key] = parseFloat(current.value);
                return;
            case ParameterTypes.Text:
                extracted[key] = current.value;
                return;
            case ParameterTypes.SharedFileSystem:
                console.log(current.value);
                extracted[key] = {fileSystemId: current.value};
                return;
            case ParameterTypes.Peer:
                extracted[key] = {jobId: current.value};
                return;
        }
    });
    return extracted;
}

export const inCancelableState = (state: AppState) =>
    state === AppState.VALIDATED ||
    state === AppState.PREPARED ||
    state === AppState.SCHEDULED ||
    state === AppState.RUNNING;


export function validateOptionalFields(
    invocation: ApplicationInvocationDescription,
    parameters: ParameterValues
): boolean {
    const optionalErrors = [] as string[];
    const optionalParams = invocation.parameters.filter(it => it.optional && it.visible).map(it =>
        ({name: it.name, title: it.title})
    );
    optionalParams.forEach(it => {
        const param = parameters.get(it.name)!;
        if (!param.current!.checkValidity()) optionalErrors.push(it.title);
    });

    if (optionalErrors.length > 0) {
        snackbarStore.addFailure(
            `Invalid values for ${optionalErrors.slice(0, 3).join(", ")}
                    ${optionalErrors.length > 3 ? `and ${optionalErrors.length - 3} others` : ""}`,
            5000
        );
        return false;
    }

    return true;
}

export function checkForMissingParameters(
    parameters: ExtractedParameters,
    invocation: ApplicationInvocationDescription
): boolean {
    const requiredParams = invocation.parameters.filter(it => !it.optional);
    const missingParameters: string[] = [];
    requiredParams.forEach(rParam => {
        const parameterValue = parameters[rParam.name];
        // Number, string, boolean 
        if (!parameterValue) {
            missingParameters.push(rParam.title);
        } else if (rParam.type === ParameterTypes.InputDirectory || rParam.type === ParameterTypes.InputFile) {
            if (!parameterValue["source"]) {
                missingParameters.push(rParam.title);
            }
        } else if (rParam.type === ParameterTypes.SharedFileSystem) {
            if (!parameterValue["fileSystemId"]) {
                missingParameters.push(rParam.title);
            }
        }
    });

    // Check missing values for required input fields.
    if (missingParameters.length > 0) {
        snackbarStore.addFailure(
            `Missing values for ${missingParameters.slice(0, 3).join(", ")} 
                ${missingParameters.length > 3 ? `and ${missingParameters.length - 3} others.` : ``}`,
            5000
        );
        return false;
    }
    return true;
}
