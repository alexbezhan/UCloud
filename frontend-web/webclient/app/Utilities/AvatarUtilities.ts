export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
