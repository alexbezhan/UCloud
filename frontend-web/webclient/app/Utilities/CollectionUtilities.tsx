export function removeEntry<T>(arr: T[], index: number): T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1));
}

export function setsDiffer<T>(s1: Set<T>, s2: Set<T>): boolean {
    if (s1.size !== s2.size) return true;
    let differ = false;
    s1.forEach(it => {
        if (!s2.has(it)) differ = true;
    });
    return differ;
}

type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
