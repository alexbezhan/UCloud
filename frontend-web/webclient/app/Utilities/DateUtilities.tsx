import {format} from "date-fns/esm";

// Could potentially cause issues with time if user is outside CEST
export const dateToString = (date: number) => format(date, "HH:mm dd/MM/yyyy");
