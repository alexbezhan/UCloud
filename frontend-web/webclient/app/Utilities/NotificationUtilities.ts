export const notificationsQuery = "/notifications?itemsPerPage=100";
export const readAllNotificationsQuery = "notifications/read/all";
export const readNotificationQuery = (id: number) => `notifications/read/${id}`;
