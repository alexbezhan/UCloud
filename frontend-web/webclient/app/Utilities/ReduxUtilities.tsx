import activity from "Activity/Redux/ActivityReducer";
import * as AppRedux from "Applications/Redux";
import analyses from "Applications/Redux/AnalysesReducer";
import applications from "Applications/Redux/BrowseReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import * as TaskRedux from "BackgroundTasks/redux";
import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import fileInfo from "Files/Redux/FileInfoReducer";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import uploader from "Uploader/Redux/UploaderReducer";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    hookStore,
    ...AppRedux.reducers,
    avatar,
    loading,
    tasks: TaskRedux.reducer,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
