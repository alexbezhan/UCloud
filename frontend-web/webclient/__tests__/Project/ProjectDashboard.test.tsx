import * as React from "react";
import ProjectDashboard from "Project/ProjectDashboard";
import {create} from "react-test-renderer";
import {Provider} from "react-redux";
import {store} from "Utilities/ReduxUtilities";
import {ThemeProvider} from "styled-components";
import theme from "../../app/ui-components/theme";
import {MemoryRouter} from "react-router";

jest.mock("Utilities/ProjectUtilities", () => ({
    isAdminOrPI: () => false
}));


jest.mock("Project/index", () => ({
    useProjectManagementStatus: () => ({
        projectId: "this-is-an-id",
        projectDetails: {data: {title: "this-is-a-title"}},
        projectRole: "ADMIN"
    })
}));

test("Mount ProjectDashboard", () => {
    expect(create(
        <MemoryRouter>
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <ProjectDashboard />
                </ThemeProvider>
            </Provider>
        </MemoryRouter>
    ).toJSON()).toMatchSnapshot();
});
