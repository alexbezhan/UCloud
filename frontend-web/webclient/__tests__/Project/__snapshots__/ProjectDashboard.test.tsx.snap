// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount ProjectDashboard 1`] = `
.c0 {
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c5 {
  padding-top: 96px;
  padding-right: 14px;
}

.c14 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c15 {
  color: #8393A7;
}

.c17 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 4px;
  padding-bottom: 4px;
}

.c12 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin: 8px;
  margin-left: 0;
  color: #53657d;
}

.c16 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  color: #53657d;
}

.c10 {
  width: 100%;
  height: auto;
  overflow: hidden;
  height: auto;
  box-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,.14),0px 1px 8px 0px rgba(0,0,0,.12);
  border: 0px solid #c9d3df;
  border-radius: 6px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: grid;
  margin-top: 2px;
  width: 100%;
  grid-gap: 10px;
  grid-template-columns: repeat(auto-fill,minmax(300px,1fr) );
}

.c13 {
  cursor: inherit;
  font-size: 24px;
  margin: 0;
  font-weight: 400;
}

.c23 {
  cursor: inherit;
  font-size: 14px;
  color: #53657d;
}

.c4 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #1e252e;
}

.c4:hover {
  color: var(--textHighlight);
}

.c1 {
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 204px;
  padding-right: 14px;
  width: 100%;
  height: 96px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

.c18 {
  background-color: #fff;
  border: 0px;
  border-spacing: 0;
  table-layout: fixed;
  width: 100%;
  min-width: 15em;
  background-color: #fff;
}

.c21 {
  border: 0px;
  border-spacing: 0;
}

.c22 {
  border: 0px;
  border-spacing: 0;
  text-align: right;
}

.c19 {
  cursor: pointer;
}

.c19 > .c20 {
  border-spacing: 0;
  border-top: 1px solid rgba(34,36,38,.1);
  padding-top: 8px;
  padding-bottom: 8px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 12px;
  width: calc(100% - 180px);
  max-width: unset;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 > span {
  width: 1;
  font-size: 25px;
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c3 > span + span:before {
  padding: 0 8px;
  vertical-align: top;
  color: var(--text,#f00);
  content: "/";
}

.c3 > span {
  color: var(--text,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3 > span:hover {
  cursor: pointer;
  color: var(--blue,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3 > span:last-child:hover {
  color: var(--text,#f00);
  cursor: default;
}

.c3 .sc-AxjAm {
  margin: 0;
  display: inline;
}

.c8 > .c9 {
  position: relative;
  min-height: 200px;
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c8 > .c9:hover {
  -webkit-transform: scale(1.02);
  -ms-transform: scale(1.02);
  transform: scale(1.02);
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c6 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c6 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c6 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="sc-AxiKw c1"
    height={96}
    width={1}
  >
    <div
      className="c2"
    >
      <div
        className="sc-fzpdyU c3"
      >
        <span>
          <a
            className="c4"
            color="text"
            href="/projects"
            onClick={[Function]}
          >
            My Projects
          </a>
        </span>
        <span>
          <a
            className="c4"
            color="text"
            href="/project/dashboard"
            onClick={[Function]}
          >
            this-is-a-title
            
          </a>
        </span>
      </div>
    </div>
  </div>
  <div
    className="c5"
  >
    <div
      className="c6"
    />
    <div
      className="c7 c8"
      width="100%"
    >
      <div
        className="c9 c10"
        height="auto"
        onClick={[Function]}
        overflow="hidden"
        width={1}
      >
        <div
          className="c11"
          style={
            Object {
              "borderTop": "5px solid var(--blue, #f00)",
            }
          }
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color="#53657d"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="20"
            m={8}
            ml={0}
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M9.293 14.795a1.848 1.848 0 00-.278-.649l-.15-.228c-.835-.959-1.292-2.353-1.534-4.07l.018-.008c.198-.077.386-.157.57-.24 4.687-2.088 5.765-5.764 5.765-5.764s1.238 2.082 2.115 4.086l.08.185.008.02c-.153 2.526-.607 4.53-1.705 5.791l-.15.229a1.84 1.84 0 00-.281.661c-.041.225-.055.45-.055.674a30.352 30.352 0 005.029-1.231c0-17.284-7.202-14.002-7.202-14.002S4.322-3.033 4.322 14.251a30.337 30.337 0 005.03 1.23 3.838 3.838 0 00-.059-.686zM21.4 19.743c-.927-.795-3.649-1.372-5.472-1.683-.139 2.322-2.047 4.168-4.405 4.168s-4.265-1.846-4.405-4.168c-1.823.311-4.545.888-5.472 1.683C.207 20.98 0 24 0 24h23.045s-.207-3.02-1.646-4.257z"
              fillRule="nonzero"
            />
            <path
              d="M11.523 21.341c-1.862 0-3.408-1.464-3.52-3.334l-.006-.088c.528-.08.867-.123.867-.123s.496-1.094.488-2.314a3.838 3.838 0 00-.059-.687 1.848 1.848 0 00-.278-.649l-.15-.228c-.835-.959-1.292-2.353-1.534-4.07l.018-.008c.198-.077.386-.157.57-.24 4.687-2.088 5.765-5.764 5.765-5.764s1.238 2.082 2.115 4.086l.08.185.008.02c-.153 2.526-.607 4.53-1.705 5.791l-.15.229a1.84 1.84 0 00-.281.661c-.041.225-.055.45-.055.674-.008 1.22.486 2.314.486 2.314s.34.043.867.123l-.005.088c-.113 1.87-1.659 3.334-3.52 3.334z"
              fill="currentcolor"
              fillRule="nonzero"
            />
          </svg>
          <h3
            className="c13"
            cursor="inherit"
            fontSize={4}
          >
            Members
          </h3>
          <div
            className="c14"
          />
          <div
            className="c15"
            color="#8393A7"
          >
            <svg
              className="c16"
              clipRule="evenodd"
              color="darkGray"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height={18}
              rotation={-90}
              viewBox="0 0 18 24"
              width={18}
            >
              <path
                d="M9 24l9-10h-6V0H6v14h-6l9 10z"
              />
            </svg>
          </div>
        </div>
        <div
          className="c17"
        >
          <table
            className="c18"
            width="100%"
          >
            <tbody>
              <tr
                className="c19"
                cursor="pointer"
              >
                <td
                  className="c20 c21"
                >
                  Members
                </td>
                <td
                  className="c20 c22"
                >
                  0
                </td>
              </tr>
              <tr
                className="c19"
                cursor="pointer"
              >
                <td
                  className="c20 c21"
                >
                  Groups
                </td>
                <td
                  className="c20 c22"
                >
                  0
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        className="c9 c10"
        height="auto"
        onClick={[Function]}
        overflow="hidden"
        width={1}
      >
        <div
          className="c11"
          style={
            Object {
              "borderTop": "5px solid var(--purple, #f00)",
            }
          }
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color="#53657d"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="20"
            m={8}
            ml={0}
            viewBox="0 0 21 24"
            width="20"
          >
            <path
              d="M10.394-.002l10.392 6v12l-10.392 6-10.392-6v-12l10.392-6z"
            />
            <path
              d="M10.393 6l5.196 3v6l-5.196 3-5.196-3V9l5.196-3z"
              fill="currentcolor"
            />
          </svg>
          <h3
            className="c13"
            cursor="inherit"
            fontSize={4}
          >
            Resource Allocation
          </h3>
          <div
            className="c14"
          />
          <div
            className="c15"
            color="#8393A7"
          >
            <svg
              className="c16"
              clipRule="evenodd"
              color="darkGray"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height={18}
              rotation={-90}
              viewBox="0 0 18 24"
              width={18}
            >
              <path
                d="M9 24l9-10h-6V0H6v14h-6l9 10z"
              />
            </svg>
          </div>
        </div>
        <div
          className="c17"
        />
      </div>
      <div
        className="c9 c10"
        height="auto"
        onClick={[Function]}
        overflow="hidden"
        width={1}
      >
        <div
          className="c11"
          style={
            Object {
              "borderTop": "5px solid var(--green, #f00)",
            }
          }
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color="#53657d"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="20"
            m={8}
            ml={0}
            viewBox="0 0 16 24"
            width="20"
          >
            <path
              d="M1.501 0a1.5 1.5 0 00-1.5 1.5v4.718A2 2 0 00.647 7.69L5.332 12 .645 16.32a2 2 0 00-.644 1.47V22.5a1.5 1.5 0 001.5 1.5h13a1.5 1.5 0 001.5-1.5V17.79a2 2 0 00-.644-1.47L10.67 12l4.685-4.31a2 2 0 00.646-1.472V1.5a1.5 1.5 0 00-1.5-1.5h-13zm11.831 17.532v4.125H2.67v-4.125L8 12.617l5.331 4.915zm-5.331-6.144L2.67 6.473V2.284h10.668v4.183l-5.337 4.921z"
              fillRule="nonzero"
            />
            <path
              d="M2.67 6.472l10.669-.006L8 11.388 2.67 6.472zM2.67 18.533h10.662v3.125H2.67z"
              fill="currentcolor"
            />
            <path
              d="M8 12.618l1 6.915H7l1-6.915z"
              fill="currentcolor"
            />
          </svg>
          <h3
            className="c13"
            cursor="inherit"
            fontSize={4}
          >
            Usage
          </h3>
          <div
            className="c14"
          />
          <div
            className="c15"
            color="#8393A7"
          >
            <svg
              className="c16"
              clipRule="evenodd"
              color="darkGray"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height={18}
              rotation={-90}
              viewBox="0 0 18 24"
              width={18}
            >
              <path
                d="M9 24l9-10h-6V0H6v14h-6l9 10z"
              />
            </svg>
          </div>
        </div>
        <div
          className="c17"
        >
          <div
            className="c23"
            color="darkGray"
            cursor="inherit"
            fontSize={1}
          >
            Past 30 days
          </div>
          <table
            className="c18"
            width="100%"
          >
            <tbody>
              <tr
                className="c19"
                cursor="pointer"
              >
                <td
                  className="c20 c21"
                >
                  Storage
                </td>
                <td
                  className="c20 c22"
                >
                  0 DKK
                </td>
              </tr>
              <tr
                className="c19"
                cursor="pointer"
              >
                <td
                  className="c20 c21"
                >
                  Compute
                </td>
                <td
                  className="c20 c22"
                >
                  0 DKK
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
